### PR TITLE
add dealine to close main server info fetch stream for unary rpcs.

### PR DESCRIPTION
**Pull request by abb3r**
Signed-off-by: Rahman Abber Tahir [rahman.abber.tahir@ibm.com](mailto:rahman.abber.tahir@ibm.com)

### CODE REVIEW

**Comment by abb3r** [comment]
Now when we have such hangs we would see:

gwhisper --rpcTimeoutMilliseconds=2000 127.0.0.1 examples.StatusHandling rpcSleepForSeconds number=5
RPC failed ;( Status code: 4 DEADLINE_EXCEEDED, error message: Deadline Exceeded
Note: You can increase the deadline by setting the --rpcTimeoutMilliseconds option to a number or 'None'.
ServerReflectionInfo rpc failed. Grpc Server failed to close the stream within 10 seconds.
Error code: 4, message: Deadline Exceeded, debug info: {"created":"@1699964194.163491587","description":"Error received from peer ipv4:127.0.0.1:50051","file":"/home/rahman/repos/os-gwhisper/gWhisper/build/_deps/grpc-src/src/core/lib/surface/call.cc","file_line":1075,"grpc_message":"Deadline Exceeded","grpc_status":4}

**Comment by abb3r** [comment]
Hi @rainerschoe, as we discussed about the 'gwhisper hangs when server fails to close the reflection stream'
I think this change would be a good enough circumvention since we only find issue with unary rpcs and they always have a default timeout of 10seconds. This would provide a safety net and ease up test scenarios until there is a change for a more dynamic timeout setting, what do you say?

File: third_party/gRPC_utils/proto_reflection_descriptor_database.cc [hunk 1]
```diff
@@ -315,7 +326,11 @@ void ProtoReflectionDescriptorDatabase::AddFileFromResponse(
 
 const std::shared_ptr<ProtoReflectionDescriptorDatabase::ClientStream>
 ProtoReflectionDescriptorDatabase::GetStream() {
-  if (!stream_) {
+  if (!stream_) //only assign deadline to a unary rpc.
```
Comment (rainerschoe) [inline third_party/gRPC_utils/proto_reflection_descriptor_database.cc:329]: I think this is not the place you are looking for.
It is kind of a stream singleton constructor. if stream does not exist it is creating it.
reflection is always a stream.

File: third_party/gRPC_utils/proto_reflection_descriptor_database.cc [hunk 2]
```diff
@@ -31,6 +31,7 @@ using grpc::reflection::v1alpha::ServerReflection;
 using grpc::reflection::v1alpha::ServerReflectionRequest;
 using grpc::reflection::v1alpha::ServerReflectionResponse;
 
+const uint8_t g_timeoutGrpcMainStreamSeconds = 10; //using default gwhisper timeout of 10 seconds.
```
Comment (rainerschoe) [inline third_party/gRPC_utils/proto_reflection_descriptor_database.cc:34]: I have a new idea:
gWhisper uses reflection only BEFORE running the actual user-requested RPC:

1. gWhisper starts parsing CLI arguments until hostname
2. once hostname is known, reflection stream is established with the server and via multiple requests all the RPC names and type information is fetched from the reflection endpoint while CLI args are parsed and verified against this.
3. The parsed CLI args are converted into a protobuf message
4. The actual RPC is called as requested with the user (may be streaming or unary RPC)
5. response is formatted and printed

I would need to verify, but I guess that currently, we keep the reflection stream open for the whole duration. which causes problems when using a fixed reflection deadline.

What we could do is close the reflection stream after 2 or 3. This way we could set a fixed reflection stream RPC timeout and we would not influence 4, even if the user executes a very long-running RPC.

**Review by rainerschoe — COMMENTED** [review]
Thanks for the PR (even if it is closed now). See my comment on your default timeout. This could actually work, when we close the reflection stream early enough.

File: third_party/gRPC_utils/proto_reflection_descriptor_database.cc (hunk 2 above)
Comment (abb3r) [reply third_party/gRPC_utils/proto_reflection_descriptor_database.cc:34]: Hi @rainerschoe thanks for the review.  I did see some deficiencies via functional tests and I like your idea of closing the desc stream early.

Hence, in the new PR
- Dynamic deadlines are now propagated to the DescDb stream.
- The db stream stream is now closed after 2.

**Pull request closed without merging**